File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/service/PersianCalendarTileService.java
package com.byagowi.persiancalendar.service;

import android.annotation.TargetApi;
import android.content.Intent;
import android.graphics.drawable.Icon;
import android.os.Build;
import android.service.quicksettings.Tile;
import android.service.quicksettings.TileService;
import android.util.Log;

import com.byagowi.persiancalendar.calendar.AbstractDate;
import com.byagowi.persiancalendar.ui.MainActivity;
import com.byagowi.persiancalendar.utils.Utils;

/**
 * Created by Alireza Afkar on 19/6/2018AD.
 */
@TargetApi(Build.VERSION_CODES.N)
public class PersianCalendarTileService extends TileService {
    @Override
    public void onClick() {
        try {
            startActivityAndCollapse(new Intent(this, MainActivity.class)
                    .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK));
        } catch (Exception e) {
            Log.e("TileService", "Tile onClick fail", e);
        }
    }

    @Override
    public void onStartListening() {
        Tile tile = getQsTile();
        if (tile == null) return;
        AbstractDate today = Utils.getTodayOfCalendar(Utils.getMainCalendar());

        tile.setIcon(Icon.createWithResource(this,
                Utils.getDayIconResource(today.getDayOfMonth())));
        tile.setLabel(Utils.getWeekDayName(today));
        tile.setContentDescription(Utils.getMonthName(today));
        // explicitly set Tile state to Active, fixes tile not being lit on some Samsung devices
        tile.setState(Tile.STATE_ACTIVE);
        tile.updateTile();
    }
}

File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/service/PersianCalendarTileService.kt
package com.byagowi.persiancalendar.service

import android.annotation.TargetApi
import android.content.Intent
import android.graphics.drawable.Icon
import android.os.Build
import android.service.quicksettings.Tile
import android.service.quicksettings.TileService
import com.byagowi.persiancalendar.ui.MainActivity
import com.byagowi.persiancalendar.utils.Jdn
import com.byagowi.persiancalendar.utils.createStatusIcon
import com.byagowi.persiancalendar.utils.dayOfWeekName
import com.byagowi.persiancalendar.utils.getDayIconResource
import com.byagowi.persiancalendar.utils.logException
import com.byagowi.persiancalendar.utils.mainCalendar
import com.byagowi.persiancalendar.utils.monthName

/**
 * Created by Alireza Afkar on 19/6/2018AD.
 */
@TargetApi(Build.VERSION_CODES.N)
class PersianCalendarTileService : TileService() {

    override fun onClick() = runCatching {
        startActivityAndCollapse(
            Intent(this, MainActivity::class.java).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        )
    }.onFailure(logException).let {}

    override fun onStartListening() = runCatching {
        qsTile?.also { tile ->
            val jdn = Jdn.today
            val today = jdn.toCalendar(mainCalendar)
            if ((false)) { // Maybe increases quick tile start time? It already gives ANRs so we should be careful
                tile.icon = Icon.createWithBitmap(createStatusIcon(this, today.dayOfMonth))
            } else {
                tile.icon = Icon.createWithResource(this, getDayIconResource(today.dayOfMonth))
            }
            tile.label = jdn.dayOfWeekName
            tile.contentDescription = today.monthName
            // explicitly set Tile state to Active, fixes tile not being lit on some Samsung devices
            tile.state = Tile.STATE_ACTIVE
        }?.updateTile()
    }.onFailure(logException).let {}
}
